ext/intl/dateformat/datetimepatterngenerator_methods.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Mel Dafert (mel@dafert.at)                                  |
+   +----------------------------------------------------------------------+
+*/
+
+#include "../intl_cppshims.h"
+
+#include <unicode/dtptngen.h>
+
+#include "../intl_convertcpp.h"
+
+extern "C" {
+#include "php_intl.h"
+#define USE_DATETIMEPATTERNGENERATOR_POINTER 1
+#include "datetimepatterngenerator_class.h"
+#include <zend_exceptions.h>
+#include <assert.h>
+}
+
+using icu::DateTimePatternGenerator;
+using icu::Locale;
+using icu::StringPiece;
+
+/* {{{ Internal constructor for IntlDateTimePatternGenerator */
+static zend_result dtpg_ctor(INTERNAL_FUNCTION_PARAMETERS)
+{
+	zval		*object;
+	char		*locale_str;
+	size_t		locale_len	= 0;
+	Locale		locale;
+	IntlDateTimePatternGenerator_object* dtpgo;
+
+	intl_error_reset(NULL);
+	object = return_value;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STRING_OR_NULL(locale_str, locale_len)
+	ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
+
+	DTPATTERNGEN_METHOD_FETCH_OBJECT_NO_CHECK;
+
+	if (dtpgo->dtpg != NULL) {
+		intl_errors_set(DTPATTERNGEN_ERROR_P(dtpgo), U_ILLEGAL_ARGUMENT_ERROR, "datetimepatterngenerator_create: cannot call constructor twice", 0);
```

Comment:
What's the reason of this? I see that the check is from `IntlDateFormatter`, but I don't know why these are necessary.

Reply:
For `IntlDateFormatter`, there is this test:
https://github.com/php/php-src/blob/8d743d5281c29e9750e183804b7ba02e1ff82f0b/ext/intl/tests/bug62081.phpt#L12-L13
How is that (anti-)pattern handled in other cases?